### PR TITLE
[openshift_manageiq] Adding module calls instead of command for idempotency.

### PR DESCRIPTION
**Pull request by kwoodson**

### CODE REVIEW

**Comment by kwoodson** [comment]
@simon3z, I'm not sure I am able to test these.  If there is an easy way to do so then please point me in the right direction.  Do you have someone who could test these changes?

**Comment by kwoodson** [comment]
Trello card here: https://trello.com/c/ZPIwuaDr/373-2-idempotency-configure-role-user-permissions

**Comment by simon3z** [comment]
@enoodle @cben can you review/test this?

**Comment by mtnbikenc** [comment]
@kwoodson I'll try to test this one as well if I have time.  Might want to go ahead and fix the yamllint.

**Comment by kwoodson** [comment]
@mtnbikenc, good call.  I didn't see those.  I repushed with fixes.

**Comment by kwoodson** [comment]
aos-ci-test

**Comment by simon3z** [comment]
@ilackarms can you check if this is OK. Thanks.

**Comment by tbielawa** [comment]
testing now

**Comment by ilackarms** [comment]
@simon3z i'm not currently using openshift-ansible; it will take me some time to setup an environment. i'll keep you updated when i've started testing this change

**Comment by mtnbikenc** [comment]
Testing build with `openshift_use_manageiq=true`

**Comment by mtnbikenc** [comment]
Restarting tests with a fresh set of VMs.

File: roles/openshift_manageiq/tasks/main.yaml [hunk 1]
```diff
@@ -3,76 +3,58 @@
...
-    echo {{ manageiq_cluster_role | to_json | quote }} |
-    {{ openshift.common.client_binary }} create
-    --config={{manage_iq_tmp_conf}}
-    -f -
-  register: osmiq_create_cluster_role
-  failed_when: "'already exists' not in osmiq_create_cluster_role.stderr and osmiq_create_cluster_role.rc != 0"
-  changed_when: osmiq_create_cluster_role.rc == 0
+- name: Create manageiq cluster role
+  oc_clusterrole:
+    name: management-infra-admin
```
Comment (mtnbikenc) [inline roles/openshift_manageiq/tasks/main.yaml]: Failure on this task:
```
2017-04-12 08:45:25,570 p=24142 u=rteague |  TASK [openshift_manageiq : Create manageiq cluster role] ***********************
2017-04-12 08:45:25,570 p=24142 u=rteague |  task path: /home/rteague/dev/clusters/aws-c1/openshift-ansible/roles/openshift_manageiq/tasks/main.yaml:20
2017-04-12 08:45:27,445 p=24142 u=rteague |  changed: [ec2-54-198-186-155.compute-1.amazonaws.com] => {
    "changed": true, 
    "results": {
        "cmd": "/bin/oc get clusterrole management-infra-admin -o json", 
        "results": [
            {}
        ], 
        "returncode": 0, 
        "stderr": "Error from server: clusterrole \"management-infra-admin\" not found\n", 
        "stdout": ""
    }, 
    "state": "present"
}
```

File: roles/openshift_manageiq/tasks/main.yaml [hunk 2]
```diff
@@ -3,76 +3,58 @@
...
-    --config={{manage_iq_tmp_conf}}
-    create -f -
-  register: oshawkular_create_cluster_role
-  failed_when: "'already exists' not in oshawkular_create_cluster_role.stderr and oshawkular_create_cluster_role.rc != 0"
-  changed_when: oshawkular_create_cluster_role.rc == 0
-  # AUDIT:changed_when_note: Checking the return code is insufficient
-  # here. We really need to verify the if the role even exists before
-  # we run this task.
+  oc_clusterrole:
+    name: hawkular-metrics-admin
```
Comment (mtnbikenc) [inline roles/openshift_manageiq/tasks/main.yaml]: Task failure:
```
2017-04-12 08:45:27,478 p=24142 u=rteague |  TASK [openshift_manageiq : Create Hawkular Metrics Admin Cluster Role] *********
2017-04-12 08:45:27,478 p=24142 u=rteague |  task path: /home/rteague/dev/clusters/aws-c1/openshift-ansible/roles/openshift_manageiq/tasks/main.yaml:32
2017-04-12 08:45:29,062 p=24142 u=rteague |  changed: [ec2-54-198-186-155.compute-1.amazonaws.com] => {
    "changed": true, 
    "results": {
        "cmd": "/bin/oc get clusterrole hawkular-metrics-admin -o json", 
        "results": [
            {}
        ], 
        "returncode": 0, 
        "stderr": "Error from server: clusterrole \"hawkular-metrics-admin\" not found\n", 
        "stdout": ""
    }, 
    "state": "present"
}
```

Comment (kwoodson) [inline roles/openshift_manageiq/tasks/main.yaml]: @mtnbikenc, I updated the module and pushed a couple of changes to the oc_clusterrole call.

File: roles/openshift_manageiq/tasks/main.yaml (hunk 1 above)
Comment (kwoodson) [inline roles/openshift_manageiq/tasks/main.yaml]: I ran this test on my cluster in its own playbook and didn't see the same behavior.

```yaml
  tasks:

  - name: Create manageiq cluster role
    oc_clusterrole:
      name: management-infra-admin
      rules:
      - apiGroups:
        - ""
        resources:
        - pods/proxy
        verbs:
        - "*"
  - name: Create Hawkular Metrics Admin Cluster Role
    oc_clusterrole:
      name: hawkular-metrics-admin
      rules:
      - apiGroups:
        - ""
        resources:
        - hawkular-alerts
        - hawkular-metrics
        verbs:
        - "*"
```

**Comment by mtnbikenc** [comment]
Starting local tests with the new changes.

File: roles/openshift_manageiq/tasks/main.yaml [hunk 3]
```diff
@@ -3,76 +3,56 @@
...
-  with_items: "{{manage_iq_tasks}}"
-  register: osmiq_perm_task
-  failed_when: "'already exists' not in osmiq_perm_task.stderr and osmiq_perm_task.rc != 0"
-  changed_when: osmiq_perm_task.rc == 0
-  # AUDIT:changed_when_note: Checking the return code is insufficient
-  # here. We really need to compare the current role/user permissions
-  # with their expected state. I think we may have a module for this?
-
+  oc_adm_policy_user:
+    namespace: management-infra
```
Comment (mtnbikenc) [inline roles/openshift_manageiq/tasks/main.yaml]: For some reason, this task reported `changed` on the second run (for idempotency), but it was only on the first two items:
```
2017-04-12 16:41:19,707 p=28455 u=rteague |  TASK [openshift_manageiq : Configure role/user permissions] ********************
2017-04-12 16:41:19,707 p=28455 u=rteague |  task path: /home/rteague/dev/clusters/aws-c1/openshift-ansible/roles/openshift_manageiq/tasks/main.yaml:43
2017-04-12 16:41:21,015 p=28455 u=rteague |  changed: [ec2-54-91-22-173.compute-1.amazonaws.com] => (item={u'resource_kind': u'role', u'resource_name': u'admin', u'user': u'management-admin'}) => {
    "changed": true, 
    "item": {
        "resource_kind": "role", 
        "resource_name": "admin", 
        "user": "management-admin"
    }, 
    "present": "present", 
    "results": {
        "cmd": "/bin/oc adm policy add-role-to-user admin management-admin -n management-infra", 
        "results": "", 
        "returncode": 0
    }
}
2017-04-12 16:41:22,047 p=28455 u=rteague |  changed: [ec2-54-91-22-173.compute-1.amazonaws.com] => (item={u'resource_kind': u'role', u'resource_name': u'management-infra-admin', u'user': u'management-admin'}) => {
    "changed": true, 
    "item": {
        "resource_kind": "role", 
        "resource_name": "management-infra-admin", 
        "user": "management-admin"
    }, 
    "present": "present", 
    "results": {
        "cmd": "/bin/oc adm policy add-role-to-user management-infra-admin management-admin -n management-infra", 
        "results": "", 
        "returncode": 0
    }
}
2017-04-12 16:41:22,905 p=28455 u=rteague |  ok: [ec2-54-91-22-173.compute-1.amazonaws.com] => (item={u'resource_kind': u'cluster-role', u'resource_name': u'cluster-reader', u'user': u'system:serviceaccount:management-infra:management-admin'}) => {
    "changed": false, 
    "item": {
        "resource_kind": "cluster-role", 
        "resource_name": "cluster-reader", 
        "user": "system:serviceaccount:management-infra:management-admin"
    }, 
    "present": "present"
}
2017-04-12 16:41:23,730 p=28455 u=rteague |  ok: [ec2-54-91-22-173.compute-1.amazonaws.com] => (item={u'resource_kind': u'scc', u'resource_name': u'privileged', u'user': u'system:serviceaccount:management-infra:management-admin'}) => {
    "changed": false, 
    "item": {
        "resource_kind": "scc", 
        "resource_name": "privileged", 
        "user": "system:serviceaccount:management-infra:management-admin"
    }, 
    "present": "present"
}
2017-04-12 16:41:24,573 p=28455 u=rteague |  ok: [ec2-54-91-22-173.compute-1.amazonaws.com] => (item={u'resource_kind': u'cluster-role', u'resource_name': u'system:image-puller', u'user': u'system:serviceaccount:management-infra:inspector-admin'}) => {
    "changed": false, 
    "item": {
        "resource_kind": "cluster-role", 
        "resource_name": "system:image-puller", 
        "user": "system:serviceaccount:management-infra:inspector-admin"
    }, 
    "present": "present"
}
2017-04-12 16:41:25,397 p=28455 u=rteague |  ok: [ec2-54-91-22-173.compute-1.amazonaws.com] => (item={u'resource_kind': u'scc', u'resource_name': u'privileged', u'user': u'system:serviceaccount:management-infra:inspector-admin'}) => {
    "changed": false, 
    "item": {
        "resource_kind": "scc", 
        "resource_name": "privileged", 
        "user": "system:serviceaccount:management-infra:inspector-admin"
    }, 
    "present": "present"
}
2017-04-12 16:41:26,211 p=28455 u=rteague |  ok: [ec2-54-91-22-173.compute-1.amazonaws.com] => (item={u'resource_kind': u'cluster-role', u'resource_name': u'self-provisioner', u'user': u'system:serviceaccount:management-infra:management-admin'}) => {
    "changed": false, 
    "item": {
        "resource_kind": "cluster-role", 
        "resource_name": "self-provisioner", 
        "user": "system:serviceaccount:management-infra:management-admin"
    }, 
    "present": "present"
}
2017-04-12 16:41:27,039 p=28455 u=rteague |  ok: [ec2-54-91-22-173.compute-1.amazonaws.com] => (item={u'resource_kind': u'cluster-role', u'resource_name': u'hawkular-metrics-admin', u'user': u'system:serviceaccount:management-infra:management-admin'}) => {
    "changed": false, 
    "item": {
        "resource_kind": "cluster-role", 
        "resource_name": "hawkular-metrics-admin", 
        "user": "system:serviceaccount:management-infra:management-admin"
    }, 
    "present": "present"
}
```

Comment (kwoodson) [inline roles/openshift_manageiq/tasks/main.yaml]: @mtnbikenc, thanks for testing.  Let me see if I can tell what's happening.

Comment (kwoodson) [inline roles/openshift_manageiq/tasks/main.yaml]: @mtnbikenc, I have pushed a fix for the changed state.  When performing a `add-role-to-user` call the proper place to look for the object that gets modified is in `policybindings`.  When performing a `add-cluster-role-to-user` it creates a `clusterpolicybinding`.  I had them both performing a search into the `clusterpolicybinding` list.  This is incorrect.  I have updated it to look in the proper `policybindings`.

**Comment by mtnbikenc** [comment]
@kwoodson Testing in progress.

**Comment by mtnbikenc** [comment]
aos-ci-test

**Comment by kwoodson** [comment]
[merge]

**Comment by kwoodson** [comment]
```
fatal: [172.18.11.201]: UNREACHABLE! => {
    "changed": false, 
    "generated_timestamp": "2017-04-17 14:37:16.713866", 
    "msg": "Failed to connect to the host via ssh: Warning: Permanently added '172.18.11.201' (ECDSA) to the list of known hosts.\r\nPermission denied (publickey,gssapi-keyex,gssapi-with-mic).\r\n", 
    "unreachable": true
}

```

**Comment by kwoodson** [comment]
[merge]

**Comment by kwoodson** [comment]
flake https://github.com/openshift/origin/issues/13516

**Comment by kwoodson** [comment]
[merge]

**Comment by kwoodson** [comment]
flake openshift/origin#13516
[merge]

**Comment by kwoodson** [comment]
flake https://github.com/openshift/origin/issues/13516

**Comment by sdodson** [comment]
That one doesn't have the label kind/test-flake, this one does.
flake https://github.com/openshift/origin/issues/13386

**Comment by kwoodson** [comment]
flake https://github.com/openshift/origin/issues/13386

**Comment by kwoodson** [comment]
same issue as before, `excluder`.  retrying [merge]

**Comment by kwoodson** [comment]
flake https://github.com/openshift/origin/issues/13271

**Comment by kwoodson** [comment]
[merge]

**Comment by kwoodson** [comment]
Squashed the latest commits.  Hopefully this time we can get through the bot.

aos-ci-test

**Comment by kwoodson** [comment]
@sdodson 

```
TASK [openshift_common : Install the base package for versioning] **************
task path: /usr/share/ansible/openshift-ansible/roles/openshift_common/tasks/main.yml:44
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "generated_timestamp": "2017-04-19 15:37:23.182487", 
    "msg": "No package matching 'origin-1.5.0-0.7.rc1' found available, installed or updated", 
    "rc": 126, 
    "results": [
        "No package matching 'origin-1.5.0-0.7.rc1' found available, installed or updated"
```

**Comment by sdodson** [comment]
flake on https://github.com/openshift/origin/issues/13814

**Comment by kwoodson** [comment]
@sdodson, thanks, just found the flake but you beat me to it.  https://github.com/openshift/origin/issues/13814

**Comment by sdodson** [comment]
once aos ci tests are green just merge this

**Comment by kwoodson** [comment]
aos-ci-test

**Comment by kwoodson** [comment]
[test]

**Comment by kwoodson** [comment]
[merge]

**Comment by kwoodson** [comment]
device mapper flake (does not always cause a test failure) https://github.com/openshift/origin/issues/13271

**Comment by kwoodson** [comment]
At @sdodson's `okay`, merging.

**Comment by kwoodson** [comment]
[merge]

**Comment by kwoodson** [comment]
```
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "generated_timestamp": "2017-04-20 18:17:33.321138", 
    "msg": "Available origin-docker-excluder version 3.6.0 is higher than the upgrade target version 1.5"
}
```

**Pull request merged**